best-practices.md

```diff
@@ -0,0 +1,183 @@
+# Best practices for implementing CRaC support in your application/library
+
+This guide assumes you are already familiar with the concepts and `Resource` API; please check out the [step-by-step guide](STEP-BY-STEP.md) for those. You should also already know what components need to handle the checkpoint; finding that out is described in another guide (TODO).
+
+Applications and its components are often with a simple lifecycle in mind; the application boots, then it is actively used, and in the end it enters shutdown and finally ends up in a terminated state, unable to start back - if the application needs the functionality again the component is re-created. This allows simpler reasoning and some performance optimizations by making fields final, or not protecting the access to an uninitialized component as the developer knows that it is not published yet.
+
+While usually most of the application can stay as-is, CRaC complicates the lifecycle of some components by adding the transition from active to a suspended state and back. In the suspended state, until the whole VM is terminated or before the component is restored, the rest of that application is still running and could access the component - e.g. a pool of network connections - and would find this component to be unusable at that moment. One solution is to block the thread and unblock it when the component is ready for serving again.
+
+The implementation of this synchronization depends mostly on the threading model of the application. We will refer to the synchronized component as *resource*, even though it might not implement the `Resource` interface directly.
+
+## Generic case: unknown number of threads arriving randomly
+
+The most complicated case is when we don't have any guarantees about who's calling into the resource. In order to block any access to that we will use the `java.util.concurrent.ReadWriteLock`:
+
+```java
+public class ConnectionPool implements Resource {
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+    private final Lock readLock = lock.readLock();
+    private final Lock writeLock = lock.writeLock();
+
+    /* Constructor registers this Resource */
+
+    public Connection getConnection() {
+        readLock.lock();
+        try {
+            /* actual code fetching the connection */
```

Comment:
Returning the Connection from the method cannot guarantee the wrte lock will be aqcuired when connections are done. The example suggests beforeCheckpoint can handle all Connections (keeping track of them?..) and reintialize them before users could access them. But there is no synchronization between Connection and e.g. afterRestore preventing the access to Connection